From linux-ext4-owner@vger.kernel.org Sat Mar 15 21:07:17 2008
Return-Path: <linux-ext4-owner@vger.kernel.org>
Received: from imap.linux.ibm.com ([unix socket]) by imap.linux.ibm.com
	(Cyrus v2.3.7-Invoca-RPM-2.3.7-7) with LMTPA; Sat, 15 Mar 2008 21:07:17
	-0400
X-Sieve: CMU Sieve 2.3
Received: by imap.linux.ibm.com (Postfix, from userid 101) id 283131910144;
	Sat, 15 Mar 2008 21:07:17 -0400 (EDT)
X-Spam-TestScore: none
X-Spam-TokenSummary: Bayes not run.
X-Spam-Checker-Version: SpamAssassin 3.1.7 (2006-10-05) on
	imap.linux.ibm.com
X-Spam-Level: 
X-Spam-Status: No, score=0.0 required=5.0 tests=none autolearn=disabled 
	version=3.1.7
X-Spam-Relay-Country: US ** US US US ** US US US US
Received: from smtp.linux.ibm.com (smtp.linux.ibm.com [9.26.4.197]) by
	imap.linux.ibm.com (Postfix) with ESMTP id A3FCD191009F for
	<cmm@imap.linux.ibm.com>; Sat, 15 Mar 2008 21:07:13 -0400 (EDT)
Received: from localhost (localhost.localdomain [127.0.0.1]) by
	smtp.linux.ibm.com (Postfix) with ESMTP id 4FED3C0F0 for
	<cmm@linux.ibm.com>; Sat, 15 Mar 2008 21:07:13 -0400 (EDT)
X-Virus-Scanned: amavisd-new at linux.ibm.com
Received: from BLDGATE.BOULDER.IBM.COM (bldgate.boulder.ibm.com
	[9.17.210.138]) by smtp.linux.ibm.com (Postfix) with ESMTP id 07235C078 for
	<cmm@linux.ibm.com>; Sat, 15 Mar 2008 21:07:12 -0400 (EDT)
Received:  by BLDGATE.BOULDER.IBM.COM (IBM VM SMTP Level 520) via spool
	with SMTP id 5847 ; Sat, 15 Mar 2008 19:07:12 MDT
Received: by bldgate.vnet.ibm.com (xagent2 5.9.6) via xagsmtp3 with spool
	id 7240 for cmm@linux.vnet.ibm.com; Sat, 15 Mar 2008 19:07:12 -0600 (MDT)
Received: from d03relay01.boulder.ibm.com [9.17.195.226] by
	BLDGATE.BOULDER.IBM.COM (IBM VM SMTP Level 520) via TCP with ESMTP ; Sat,
	15 Mar 2008 19:07:12 MDT
Received: from d03av04.boulder.ibm.com (d03av04.boulder.ibm.com
	[9.17.195.170]) by d03relay01.boulder.ibm.com (8.13.8/8.13.8/NCO v8.7) with
	ESMTP id m2G17Ci2115122; Sat, 15 Mar 2008 19:07:12 -0600
Received: from d03av04.boulder.ibm.com (loopback [127.0.0.1]) by
	d03av04.boulder.ibm.com (8.12.11.20060308/8.13.3) with ESMTP id
	m2G17CsC009730; Sat, 15 Mar 2008 19:07:12 -0600
Received: from d03as04.boulder.ibm.com (d03as04 [9.17.195.245]) by
	d03av04.boulder.ibm.com (8.12.11.20060308/8.12.11) with ESMTP id
	m2G17Bvh009723 (version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-SHA bits=256
	verify=OK); Sat, 15 Mar 2008 19:07:11 -0600
Received: from e33.co.us.ibm.com ([9.17.249.43]) by d03as04.boulder.ibm.com
	(8.12.11.20060308/8.12.11) with ESMTP id m2G17AXZ016628
	(version=TLSv1/SSLv3 cipher=DHE-RSA-AES256-SHA bits=256 verify=OK); Sat, 15
	Mar 2008 19:07:10 -0600
Received: from vger.kernel.org (vger.kernel.org [209.132.176.167]) by
	e33.co.us.ibm.com (8.13.8/8.13.8) with ESMTP id m2G17A3C011354; Sat, 15 Mar
	2008 21:07:10 -0400
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand id
	S1752334AbYCPA7J (ORCPT <rfc822;jrs@us.ibm.com> + 2 others); Sat, 15 Mar
	2008 20:59:09 -0400
Received: (majordomo@vger.kernel.org) by vger.kernel.org id
	S1753170AbYCPA7I (ORCPT <rfc822;linux-ext4-outgoing>); Sat, 15 Mar 2008
	20:59:08 -0400
Received: from www.church-of-our-saviour.ORG ([69.25.196.31]:33229 "EHLO
	thunker.thunk.org" rhost-flags-OK-OK-OK-FAIL) by vger.kernel.org with ESMTP
	id S1752334AbYCPA7H (ORCPT <rfc822;linux-ext4@vger.kernel.org>); Sat, 15
	Mar 2008 20:59:07 -0400
Received: from root (helo=closure.thunk.org) by thunker.thunk.org with
	local-esmtp (Exim 4.50 #1 (Debian)) id 1JahEs-0007WS-D9; Sat, 15 Mar 2008
	21:00:38 -0400
Received: from tytso by closure.thunk.org with local (Exim 4.67)
	(envelope-from <tytso@mit.edu>) id 1JahDM-0006lh-7i; Sat, 15 Mar 2008
	20:59:04 -0400
From: "Theodore Ts'o" <tytso@mit.edu>
To: linux-ext4@vger.kernel.org
Cc: "Theodore Ts'o" <tytso@mit.edu>
Subject: [PATCH, RFC] jbd2: Add commit time into the commit block
Date: 	Sat, 15 Mar 2008 20:59:04 -0400
Message-Id: <1205629144-25994-1-git-send-email-tytso@mit.edu>
X-Mailer: git-send-email 1.5.4.1.144.gdfee-dirty
X-SA-Exim-Connect-IP: <locally generated>
X-SA-Exim-Mail-From: tytso@mit.edu
X-SA-Exim-Scanned: No (on thunker.thunk.org); SAEximRunCond expanded to
	false
Sender: linux-ext4-owner@vger.kernel.org
Precedence: bulk
List-ID: <linux-ext4.vger.kernel.org>
X-Mailing-List: 	linux-ext4@vger.kernel.org
X-Xagent-From: tytso@mit.edu
X-Xagent-To: cmm@linux.vnet.ibm.com
X-Xagent-Gateway: bldgate.vnet.ibm.com (XAGENTU at BLDGATE)
X-Evolution-Source: imap://cmm@imap.linux.ibm.com/
Content-Transfer-Encoding: 8bit
Mime-Version: 1.0

Carlo Wood has demonstrated that it's possible to recover deleted
files from the journal.  Something that will make this easier is if we
can put the time of the commit into commit block.

Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
---
 fs/jbd2/commit.c     |    3 +++
 include/linux/jbd2.h |    2 ++
 2 files changed, 5 insertions(+)

Index: linux-2.6.25-rc6/fs/jbd2/commit.c
===================================================================
--- linux-2.6.25-rc6.orig/fs/jbd2/commit.c	2008-03-25 11:36:37.000000000 -0700
+++ linux-2.6.25-rc6/fs/jbd2/commit.c	2008-03-25 11:50:08.000000000 -0700
@@ -112,6 +112,7 @@ static int journal_submit_commit_record(
 	struct buffer_head *bh;
 	int ret;
 	int barrier_done = 0;
+	struct timespec now = current_kernel_time();
 
 	if (is_journal_aborted(journal))
 		return 0;
@@ -126,6 +127,8 @@ static int journal_submit_commit_record(
 	tmp->h_magic = cpu_to_be32(JBD2_MAGIC_NUMBER);
 	tmp->h_blocktype = cpu_to_be32(JBD2_COMMIT_BLOCK);
 	tmp->h_sequence = cpu_to_be32(commit_transaction->t_tid);
+	tmp->h_commit_sec = cpu_to_be32(now.tv_sec);
+	tmp->h_commit_nsec = cpu_to_be32(now.tv_nsec);
 
 	if (JBD2_HAS_COMPAT_FEATURE(journal,
 				    JBD2_FEATURE_COMPAT_CHECKSUM)) {
Index: linux-2.6.25-rc6/include/linux/jbd2.h
===================================================================
--- linux-2.6.25-rc6.orig/include/linux/jbd2.h	2008-03-16 16:32:14.000000000 -0700
+++ linux-2.6.25-rc6/include/linux/jbd2.h	2008-03-25 11:46:46.000000000 -0700
@@ -170,6 +170,8 @@ struct commit_header {
 	unsigned char   h_chksum_size;
 	unsigned char 	h_padding[2];
 	__be32 		h_chksum[JBD2_CHECKSUM_BYTES];
+	__be32		h_commit_sec;
+	__be32		h_commit_nsec;
 };
 
 /*
